src/policy_evaluator_builder.rs

```diff
@@ -74,41 +100,223 @@ impl PolicyEvaluatorBuilder {
+                    let mut wasmtime_config = wasmtime::Config::new();
+                    if self.wasmtime_cache {
+                        wasmtime_config.cache_config_load_default()?;
+                    }
+
+                    wasmtime::Engine::new(&wasmtime_config)
+                },
+                |e| Ok(e.clone()),
+            )
+            .map_err(|e| anyhow!("cannot create wasmtime engine: {:?}", e))?;
+
+        let module: wasmtime::Module = if let Some(m) = &self.policy_module {
+            // it's fine to clone a Module, this is a cheap operation that just
+            // copies its internal reference. See wasmtime docs
+            m.clone()
         } else {
-            self.policy_contents.unwrap()
+            match &self.policy_file {
+                Some(file) => wasmtime::Module::from_file(&engine, file),
+                None => wasmtime::Module::new(&engine, self.policy_contents.as_ref().unwrap()),
+            }?
         };
 
-        let mode = self
-            .execution_mode
-            .ok_or_else(|| anyhow!("Must specify execution mode"))?;
-
-        PolicyEvaluator::new(
-            self.policy_id,
-            contents,
-            mode,
-            self.settings,
-            self.callback_channel,
-            self.wasmtime_cache,
-        )
+        let execution_mode = self.execution_mode.unwrap();
+
+        let (policy, runtime) = match execution_mode {
+            PolicyExecutionMode::KubewardenWapc => {
+                let engine_provider = WasmtimeEngineProviderBuilder::new()
+                    .engine(engine)
+                    .module(module)
+                    .build()?;
+
+                let wapc_host =
+                    WapcHost::new(Box::new(engine_provider), Some(Box::new(wapc_callback)))?;
+                let policy = Self::from_contents_internal(
+                    self.policy_id.clone(),
+                    self.callback_channel.clone(),
+                    || Some(wapc_host.id()),
+                    Policy::new,
+                    execution_mode,
+                )?;
+
+                let policy_runtime = Runtime::Wapc(wapc_host);
+                (policy, policy_runtime)
+            }
+            PolicyExecutionMode::Opa | PolicyExecutionMode::OpaGatekeeper => {
+                let policy = Self::from_contents_internal(
+                    self.policy_id.clone(),
+                    self.callback_channel.clone(),
+                    || None,
+                    Policy::new,
+                    execution_mode,
+                )?;
+
+                let evaluator = burrego::Evaluator::from_engine_and_module(
+                    engine,
+                    module,
+                    crate::runtimes::burrego::new_host_callbacks(),
+                )?;
+
+                let policy_runtime = Runtime::Burrego(Box::new(BurregoEvaluator {
+                    evaluator,
+                    entrypoint_id: 0, // currently hardcoded to this value
```

Comment:
why hardcoded to 0? if we don't need it at this point should we just use the default value?

Reply:
We cannot implement the `Default` trait against `BurregoEvaluator`, there's no way to have a default `burrego::Evaluator`

We could move the hard-coded `0` into `BurregoEvaluator` here:
https://github.com/kubewarden/policy-evaluator/blob/d73d937366db297328d1cb9452b5eebe3849a2b4/src/runtimes/burrego.rs#L68-L71

However, it wouldn't be that different from the code you just commented on

Do you have other ideas?

Reply:
would hardcoding it there simplify much? AFAIU, the other runtimes have an `entrypoint_id` too?